### PR TITLE
Expose getBalance and refactor key generation

### LINKED ISSUE #7: Expose getBalance rpc method

### PR DESCRIPTION
**Pull request by MakMuftic**
Closes #7 

@mpetrunic This PR also contains refactored key generation code and export of seed instead of a private key.

### CODE REVIEW

File: test/unit/rpc/substrate/getBalance.test.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+              version: "2"
+            },
+            meta: {}
+          },
+          seed: "aba2dd1a12eeafda3fda62aa6dfa21ca",
+        }
+      }
+    } as MetamaskState);
+    // api stub
+    const apiStub = {query: {system: {account: sinon.stub()}}};
```
Comment (MakMuftic) [inline test/unit/rpc/substrate/getBalance.test.ts]: @mpetrunic I am not sure this is the best solution for stubbing API

File: src/crypto/keys.ts [hunk 2]
```diff
@@ -1,23 +1,24 @@
...
+import { cryptoWaitReady } from '@polkadot/util-crypto';
+import { stringToU8a } from '@polkadot/util';
 
 // Generate keypair from metamask wallet interface using app key
-export async function generateKeys(wallet: Wallet): Promise<KeyPairState> {
-  const appKey = await wallet.getAppKey();
-  const keypair = generateKeyPairFromSeed(appKey);
+export async function generateKeys(wallet: Wallet): Promise<KeyringPair> {
+  // get app key and wait for api to be ready
+  const result = await Promise.all([wallet.getAppKey(), cryptoWaitReady()]);
```
Comment (mpetrunic) [inline src/crypto/keys.ts]: const [appKey] = ...

File: src/crypto/keys.ts [hunk 3]
```diff
@@ -1,23 +1,24 @@
...
+  const appKey = result[0];
+  // generate keys
+  const seed = (appKey.substr(0, 32));
+  const keyring = new Keyring();
+  const keyringPair = keyring.addFromSeed(stringToU8a(seed));
+  const accountState = {keyring: keyringPair.toJson(), seed} as AccountState;
```
Comment (mpetrunic) [inline src/crypto/keys.ts]: const accountState: AccountState so you don't cast

Comment (mpetrunic) [reply src/crypto/keys.ts]: why do you store seed?
We should probably encrypt keyring with seed so attacker cannot get plaintext account from local storage

File: src/rpc/exportSeed.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  const state = wallet.getPluginState();
+  if (state != null) {
+    // ask for confirmation
+    const confirmation = await showConfirmationDialog(
+      wallet,
+      'Do you want to export your seed?'
+    );
+    // return seed if user confirmed action
+    if (confirmation) {
+      return state.polkadot.account.seed;
```
Comment (mpetrunic) [inline src/rpc/exportSeed.ts]: I think we should just use wallet.getAppKey here

File: src/rpc/getAddress.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import {Wallet} from "../interfaces";
+
+export async function getAddress(wallet: Wallet): Promise<string> {
+  const state = wallet.getPluginState();
```
Comment (mpetrunic) [inline src/rpc/getAddress.ts]: create some utility function getKeyPair which checks plugin state and generates keys if there isn't anything in state

Comment (mpetrunic) [reply src/rpc/getAddress.ts]: Use that method in every rpc that requires keys

File: src/snap.ts [hunk 6]
```diff
@@ -1,20 +1,32 @@
...
       return await getPublicKey(wallet);
+    case 'getAddress':
+      return await getAddress(wallet);
+    case 'exportSeed':
+      return await exportSeed(wallet);
+    case 'substrate_getBalance':
+      return await getBalance(wallet, api);
     case 'substrate_getChainHead':
-      const api = await initApi();
+      // temporary method
```
Comment (mpetrunic) [inline src/snap.ts]: why temporary?

File: test/unit/rpc/substrate/getBalance.test.ts (hunk 1 above)
Comment (mpetrunic) [reply test/unit/rpc/substrate/getBalance.test.ts]: I think it's fine, it's hard to stub properties

File: src/snap.ts (hunk 6 above)
Comment (MakMuftic) [reply src/snap.ts]: I used it just to test polkadot api, do we need to expose this ?

File: src/crypto/keys.ts (hunk 3 above)
Comment (MakMuftic) [reply src/crypto/keys.ts]: @mpetrunic everything else is finished, I think this should be moved to new task (**encrypt keyring**).

File: src/polkadot/getKeyPair.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import { KeyringPair } from '@polkadot/keyring/types';
+
+/**
+ * Returns KeyringPair if one is saved in wallet state, creates new one otherwise
+ * @param wallet
+ */
+export async function getKeyPair(wallet: Wallet): Promise<KeyringPair> {
+  const keyPairState = wallet.getPluginState();
+  let keyringPair: KeyringPair;
+  if (keyPairState != null) {
```
Comment (mpetrunic) [inline src/polkadot/getKeyPair.ts]: `if (!!keyPairState)` to check for undefined as well

**Pull request merged**